src/VisualStudio/Core/Impl/SolutionExplorer/AnalyzerItem/AnalyzerItemSource.cs

```diff
@@ -2,237 +2,171 @@
-                }
+        // Kick off the initial work to determine the starting set of items.
+        _workQueue.AddWork();
+    }
 
-                var sorted = _analyzerItems.OrderBy(item => item.AnalyzerReference.Display).ToArray();
-                for (var i = 0; i < sorted.Length; i++)
-                {
-                    _analyzerItems.Move(_analyzerItems.IndexOf(sorted[i]), i);
-                }
+    public object SourceItem => _analyzersFolder;
 
-                _analyzerItems.EndBulkOperation();
+    // Defer actual determination and computation of the items until later.
+    public bool HasItems => !_cancellationTokenSource.IsCancellationRequested;
 
-                NotifyPropertyChanged(nameof(HasItems));
-            }
-        }
+    public IEnumerable Items => _items;
 
-        private void NotifyPropertyChanged(string propertyName)
+    private void OnWorkspaceChanged(object sender, WorkspaceChangeEventArgs e)
+    {
+        switch (e.Kind)
         {
-            PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(propertyName));
+            case WorkspaceChangeKind.SolutionAdded:
+            case WorkspaceChangeKind.SolutionChanged:
+            case WorkspaceChangeKind.SolutionReloaded:
+            case WorkspaceChangeKind.SolutionRemoved:
+            case WorkspaceChangeKind.SolutionCleared:
+                _workQueue.AddWork();
+                break;
+
+            case WorkspaceChangeKind.ProjectAdded:
+            case WorkspaceChangeKind.ProjectReloaded:
+            case WorkspaceChangeKind.ProjectChanged:
+            case WorkspaceChangeKind.ProjectRemoved:
+                if (e.ProjectId == this.ProjectId)
+                    _workQueue.AddWork();
+
+                break;
         }
+    }
 
-        public bool HasItems
+    private async ValueTask ProcessQueueAsync(CancellationToken cancellationToken)
+    {
+        // If the project went away, then shut ourselves down.
+        var project = this.Workspace.CurrentSolution.GetProject(this.ProjectId);
+        if (project is null)
         {
-            get
-            {
-                if (_analyzerItems != null)
-                {
-                    return _analyzerItems.Count > 0;
-                }
-
-                var project = _analyzersFolder.Workspace
-                                                .CurrentSolution
-                                                .GetProject(_analyzersFolder.ProjectId);
+            this.Workspace.WorkspaceChanged -= OnWorkspaceChanged;
 
-                if (project != null)
-                {
-                    return project.AnalyzerReferences.Count > 0;
-                }
+            _cancellationTokenSource.Cancel();
 
-                return false;
-            }
+            // Note: mutating _items will be picked up automatically by clients who are bound to the collection.  We do
+            // not need to notify them through some other mechanism.
+            _items.Clear();
```

Comment:
>_items.Clear(); [](http://example.com/codeflow?start=12&length=15)

Might be good to do this not being on the ui thread too

Reply:
k

Reply:
done.